File: src/main/scala/org/joisen/chapter05/TransformMapTest.scala
package org.joisen.chapter05

import org.apache.flink.api.common.functions.MapFunction
import org.apache.flink.streaming.api.scala._

/**
 * @Author Joisen
 * @Date 2022/11/20 15:43
 * @Version 1.0
 */
object TransformMapTest {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val stream: DataStream[Event] = env.fromElements(Event("Mary", "./home", 1000L),
      Event("Bob", "./cart", 2000L))

    // 1 提取每次点击操作的用户名
    stream.map( _.user ).print("1")

    // 2 实现MapFunction接口
    stream.map(new UserExtractor).print("2")

    env.execute()
  }
  // 泛型 T：要操作的数据的类型  O：输出的类型
  class UserExtractor extends MapFunction[Event, String]{
    override def map(t: Event): String = {
      t.user
    }
  }
}

File: src/main/java/org/joisen/java/chapter05/TransformMapTest.java
package org.joisen.java.chapter05;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;

/**
 * @Author Joisen
 * @Date 2022/12/5 20:48
 * @Version 1.0
 */
public class TransformMapTest {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);
        // 进行转换计算， 提取user字段
        // 1 使用自定义类，实现MapFunction接口
        DataStreamSource<Event> stream = env.fromElements(new Event("Alice", "./home", 1000L),
                new Event("Mary", "./me", 2000L),
                new Event("Bob", "./cart", 3000L),
                new Event("Cary", "./prod?id=1", 4000L));
        // 2 使用匿名类实现MapFunction接口
        stream.map(new MapFunction<Event, String>() {

            @Override
            public String map(Event event) throws Exception {
                return event.user;
            }
        }).print();

        // 3 使用lambda表达式
        stream.map( data -> data.user).print("λ");



//        stream.map(new MyMapper()).print();


        env.execute();
    }
    // 自定义MapFunction
    public static class MyMapper implements MapFunction<Event, String>{

        @Override
        public String map(Event event) throws Exception {
            return event.user;
        }
    }

}
